src/plugins/mavlink_ftp/mavlink_ftp_impl.cpp

```diff
@@ -0,0 +1,1121 @@
+    }
+
+    close(_session_info.fd);
+    _session_info.fd = -1;
+    _session_info.stream_download = false;
+
+    payload->size = 0;
+
+    return ServerResult::SUCCESS;
+}
+
+MavlinkFTPImpl::ServerResult MavlinkFTPImpl::_work_reset(PayloadHeader* payload)
+{
+    if (_session_info.fd != -1) {
+        close(_session_info.fd);
+        _session_info.fd = -1;
+        _session_info.stream_download = false;
+    }
+
+    payload->size = 0;
+
+    return ServerResult::SUCCESS;
+}
+
+MavlinkFTPImpl::ServerResult MavlinkFTPImpl::_work_remove_directory(PayloadHeader* payload)
+{
+    std::string path = _get_path(payload);
+    if (path.rfind(_root_dir, 0) != 0) {
+        LogWarn() << "FTP: invalid path " << path;
+        return ServerResult::ERR_FAIL;
+    }
+
+    std::error_code ec;
+    if (fs::remove(path, ec) && !ec) {
+        return ServerResult::SUCCESS;
+    } else {
+        return ServerResult::ERR_FAIL;
+    }
+}
+
+MavlinkFTPImpl::ServerResult MavlinkFTPImpl::_work_create_directory(PayloadHeader* payload)
+{
+    std::string path = _get_path(payload);
+    if (path.rfind(_root_dir, 0) != 0) {
+        LogWarn() << "FTP: invalid path " << path;
+        return ServerResult::ERR_FAIL;
+    }
+
+    std::error_code ec;
+    if (fs::create_directory(path, ec) && !ec) {
+        return ServerResult::SUCCESS;
+    } else {
+        return ServerResult::ERR_FAIL;
+    }
+}
+
+MavlinkFTPImpl::ServerResult MavlinkFTPImpl::_work_remove_file(PayloadHeader* payload)
+{
+    std::string path = _get_path(payload);
+    if (path.rfind(_root_dir, 0) != 0) {
+        LogWarn() << "FTP: invalid path " << path;
+        return ServerResult::ERR_FAIL;
+    }
+
+    std::error_code ec;
+    if (fs::remove(path, ec) && !ec) {
+        return ServerResult::SUCCESS;
+    } else {
+        return ServerResult::ERR_FAIL;
+    }
+}
+
+MavlinkFTPImpl::ServerResult MavlinkFTPImpl::_work_rename(PayloadHeader* payload)
+{
+    size_t term_i = payload->size;
+    if (payload->size >= max_data_length) {
+        term_i = max_data_length - 1;
+    }
+    payload->data[term_i] = '\0';
+
+    std::string old_name = std::string(reinterpret_cast<char*>(&(payload->data[0])));
+    std::string new_name =
+        _get_path(std::string(reinterpret_cast<char*>(&(payload->data[old_name.length() + 1]))));
+    old_name = _get_path(old_name);
+    if (old_name.rfind(_root_dir, 0) != 0 || new_name.rfind(_root_dir, 0) != 0) {
+        return ServerResult::ERR_FAIL;
+    }
+    std::error_code ec;
+    fs::rename(old_name, new_name, ec);
+    if (!ec) {
+        return ServerResult::SUCCESS;
+    } else {
+        return ServerResult::ERR_FAIL;
+    }
+}
+
+MavlinkFTP::Result MavlinkFTPImpl::calc_local_file_crc32(const std::string& path, uint32_t& csum)
+{
+    crc32 checksum;
+    ssize_t bytes_read;
+    int fd = ::open(path.c_str(), O_RDONLY);
+
+    if (fd < 0) {
+        return MavlinkFTP::Result::FILE_IO_ERROR;
+    }
+
+    char buffer[18392];
```

Comment:
Why this size? What happens if the file is bigger?

Reply:
This reads the file in buffer size chunks. Code was taken from Firmware where the buffer was much smaller.

Reply:
Ok, maybe add that as a comment so it's clear for a reader.